File: src/scripts/transform.js
const fs = require('fs/promises');
const path = require('path');
const os = require('os');

const dpath = path.join(os.homedir(), '.IntelliDesk/.store')

let schema = [
    {
        metadata: {
            model: 'chat',
            name: '',
            id: '',
            created_at: '',
            update_at: '',
            highlight: ''
        },
        chats: []
    }
]

const datePattern = /^\d{2}-\d{2}-\d{2}-\d{2}-\d{2}-\d{2}$/;

function stripHtmlTags(str) {
    if (typeof str !== 'string') return '';
    return str.replace(/<[^>]*>?/g, '');
}

async function transformFn() {
    try {
        const files = await fs.readdir(dpath);
        for (let i = 0; i < files.length; i++) {
            const file = files[i]
            console.log("Processing:", i, 'Name:', file)
            const filePath = path.join(dpath, file);
            const data = await fs.readFile(filePath, 'utf-8')
            let content = JSON.parse(data)
            const name = file.split('.')[0].slice(2)
            const id = file.split('.')[0]
            let highlight = ''

            try {
                if (content[0].role === 'system') {
                    content.shift()
                }
                if (file.startsWith('C')) {
                    highlight = content[0].content
                        ?.replace(/\[\d{1,2}:\d{2} \d{2}-\d{2}-\d{4} UT\s*]?/g, '') //Remove date
                        .slice(0, 30)

                } else {
                    highlight = content[0].content[0].text
                        ?.replace(/\[\d{1,2}:\d{2} \d{2}-\d{2}-\d{4} UT\s*]?/g, '') //Remove date
                        .slice(0, 30)
                }
            } catch (err) {
                highlight = ''
            }
            highlight = (typeof highlight === 'string' && highlight.trim())
                ? highlight.trim()
                    .replace(/(\S)\n(\S)/g, '$1 $2') // Remove line breaks
                    .replace(/[\r\n]+/g, '') //Replace \n between word with space
                : '';

            schema[0].chats = content
            schema[0].metadata = {
                model: file.startsWith('C') ? 'chat' : 'multimodal',
                name: name,
                id: id,
                created_at: (name.length > 2 && datePattern.test(name)) ? name : '',
                highlight: stripHtmlTags(highlight)
            }
            //console.log(schema[0].chats)

            await fs.writeFile(filePath, JSON.stringify(schema))
        }
    } catch (err) {
        console.error('Error:', err);
    }
}

//transformFn()
//const fpath = path.join(os.homedir(), '.IntelliDesk/.store')

async function transformST() {
    const data = await fs.readFile('conversations.json', 'utf-8')
    const tdata = JSON.stringify(JSON.parse(data), null, 2)
    fs.writeFile('transformed_conversations.json', tdata);
}

//transformST()


function hasFiles(item) {
    console.log(item.key)
    const res = item?.filter(i => i?.message.files.length > 0)

    console.log(res)
    return res
}

async function transformDS() {
    const raw_data = await fs.readFile('transformed_conversations.json', 'utf-8')
    const data = JSON.parse(raw_data)
    //for (let i = 0; i < files.length; i++)
    data.forEach(conv => {
        schema[0].metadata = {
            model: hasFiles(conv.mapping) ? 'multimodal' : 'chat',
            name: '',
            id: conv.id,
            created_at: conv.inserted_at,
            highlight: conv.title,
            update_at: conv.updated_at
        }
        console.log(schema)
        return
    })
}

transformDS()

File: src/scripts/transform.py
import json
import os
import asyncio
import datetime
import time

file = 'transformed_conversations.json'

g_schema = {
    'metadata': {
        'model': 'chat',
        'name': '',
        'id': '',
        'created_at': '',
        'update_at': '',
        'highlight': ''
    },
    'chats': []
}


def hasFiles(item: dict):
    for k, v in item.items():
        for k2, v2 in v.items():
            if isinstance(v2, dict):
                if len(v2['files']) > 0:
                    return True
    return False


async def transformDS():
    with open(file, 'r') as f:
        data = json.load(f)

    count = 0

    for conv in data:
        # Create a NEW schema object for each conversation
        schema = {
            'metadata': {
                'model': 'chat',
                'name': '',
                'id': '',
                'created_at': '',
                'update_at': '',
                'highlight': ''
            },
            'chats': []  # This gets reset for each conversation
        }

        schema['metadata'] = {
            'model': 'chat',
            'name': conv['title'],
            'id': conv['id'],
            'created_at': conv['inserted_at'],
            'highlight': '',
            'update_at': conv['updated_at']
        }

        # Reset model to 'chat' for each conversation
        schema['metadata']['model'] = 'chat'

        for k, v in conv['mapping'].items():
            message = v['message']

            if not message:
                continue

            files = message.get('files', [])
            message_item = message.get('fragments', [])

            # Check for multimodal content
            if len(files) > 0:
                schema['metadata']['model'] = 'multimodal'

            for frag in message_item:
                frag_type = frag.get('type', '').lower()

                if frag_type not in ('request', 'response'):
                    continue

                _type = 'user' if frag_type == 'request' else 'assistant'

                item = {
                    'role': _type,
                    'content': frag.get('content', '')
                }

                # Check if item already exists to avoid duplication
                if not any(existing_item['role'] == item['role']
                           and existing_item['content'] == item['content']
                           for existing_item in schema['chats']):
                    schema['chats'].append(item)

                    # Set highlight only if not already set
                    if not schema['metadata']['highlight']:
                        content_words = item['content'].split()
                        cleaned_words = [
                            word.replace('\n', '').replace('<', '').replace('>', '')
                            .replace('/', '').replace('"', "'")
                            for word in content_words
                            if word.strip()
                        ]
                        highlight = cleaned_words if len(cleaned_words) <= 8 else cleaned_words[:8]
                        schema['metadata']['highlight'] = ' '.join(highlight)

        # Only save if there are chats and we have valid content
        if len(schema['chats']) > 0 and any(chat['content'].strip() for chat in schema['chats']):
            success = save(schema)
            if success:
                count += 1

    print(f"Saved: \033[32m{count}\033[0m files")
    return True


def save(data: dict):
    try:
        dir = "cache"
        name = data['metadata']['id']
        path = os.path.join(dir, f"{name}.json")

        # Create directory if it doesn't exist
        os.makedirs(dir, exist_ok=True)

        print(f"Saving: {name}")

        with open(path, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)

        return True
    except Exception as e:
        print(f"Error saving {data['metadata']['id']}: {e}")
        return False


def save_as_dict():
    dir = "/home/skye/.IntelliDesk/.store"
    for file in os.listdir(dir):
        fpath = os.path.join(dir, file)
        data = None
        try:
            with open(fpath, 'r') as f:
                data = json.load(f)
        except Exception as e:
            print(e, file)

        if data and isinstance(data, list):
            new_data = data[0]

            with open(fpath, 'w') as fw:
                json.dump(new_data, fw, indent=2)
            print(f"Saved: {fpath}")


def metadata_update():
    dir = "cache"  # "/home/skye/.IntelliDesk/.store"
    for file in os.listdir(dir):
        fpath = os.path.join(dir, file)
        data = None
        try:
            with open(fpath, 'r') as f:
                data = json.load(f)
        except Exception as e:
            print(e, file)

        o_metadata = data['metadata']

        if o_metadata.get('timestamp', None):
            data['metadata'] = {
                'model': o_metadata['model'],
                'name': o_metadata['name'],
                'id': o_metadata['id'],
                'created_at': o_metadata.get('timestamp'),
                'update_at': o_metadata['timestamp'],
                'highlight': o_metadata['highlight']
            }

            with open(fpath, 'w') as fw:
                json.dump(data, fw, indent=2)
            print(f"Updated: {fpath}")


def time_upd():
    dt = datetime.timedelta()
    print(dt)


if __name__ == "__main__":
    # save_as_dict()
    asyncio.run(transformDS())
    metadata_update()
    # time_upd()
